kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,308 @@
+
+-   Within this framing, the analyzer communicates with the driver using
+    [JSON-RPC 2.0](http://www.jsonrpc.org/specification) requests and responses,
+    as defined below. The analyzer is the JSON-RPC client, and the driver is the
+    server.
+
+    The server MUST support concurrent JSON-RPC requests from the analyzer. The
+    server MUST support batch JSON-RPC requests from the analyzer.
+
+-   Under certain circumstances, the analyzer process **must exit**. In a case
+    where the analyzer process must exit, it MUST terminate within a grace
+    period. If it does not, the operating environment SHOULD forcibly terminate
+    the process (and may do so at any time after the grace period ends). The
+    exit code of an analyzer process is ignored. The grace period is defined by
+    the implementation, but must be at least 10 seconds.
+
+-   If and after the analyzer closes the output stream, the analyzer process
+    must exit.
+
+-   If and after the driver closes the input stream, the analyzer process must
+    exit.
+
+-   A **corrupt frame** is one in which the length tag does not have the correct
+    form. A frame is corrupt if the newline terminator is missing, the string
+    preceding the newline terminator contains non-digits or is empty, or the
+    stream ends before the specified number of bytes is delivered.
+
+    If the analyzer receives a corrupt frame from the driver, it must exit. If
+    the driver receives a corrupt frame from the analyzer, it must close the
+    input stream (in which case the analyzer therefore must exit).
+
+-   When an analyzer process terminates, the driver MUST consider any pending
+    analysis requests to have failed, and thereafter the environment MAY retry
+    analysis of the corresponding compilation records.
+
+### Messages
+
+The following message structures are used by the protocol:
+
+-   `InitRequest`
+
+    ```
+    {
+      "protocol": <protocol>
+      "output-encoding": <output-encoding>
+    }
+    ```
+
+-   `InitReply`
+
+    ```
+    {
+      "protocol": <protocol>
+    }
+    ```
+
+-   `AnalyzerRequest`
+
+    ```
+    {
+      "type": <analysis-type>
+    }
+    ```
+
+-   `AnalyzeReply`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "input": <file-path>
+      "output": <file-path>
+    }
+    ```
+
+-   `LogRequest`
+
+    ```
+    {
+      "id": <analysis-id>,
+      "message": <diagnostic>
+    }
+    ```
+
+The basic types referenced above are:
+
+-   `<analysis-id>`: an integer ID assigned by the server (JSON: number).
+-   `<bytes>`: a base64-encoded byte string (JSON: string).
+-   `<diagnostic>`: a `kythe.proto.common.Diagnostic` message (JSON: object).
+-   `<digest>`: a lower-case hex-coded SHA-256 digest (JSON: string).
+-   `<file-path>`: a file path (JSON: string).
+-   `<protocol>`: a protocol version label (JSON: string).
+-   `<analysis-type>`: an identifier for a specific type or types of analyses.
```

Comment:
How are type**S** of analyses specified?  CSV?

Reply:
It should probably be a list if we want multiple values, that will make it compatible with a protobuf `repeated` field.